### PR TITLE
evolution-ews: init at 3.36.4

### PR DESCRIPTION
**Pull request by dasj19**
###### Motivation for this change
Building evolution-ews from source. Thanks to @jtojnar for helping me along the way.

###### Things done
I got evolution-ews to compile in nix.
The next steps will be to make a wrapper for evolution (evolution-with-plugins) to make use of this library.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bjornfor** [comment]
Related: https://github.com/NixOS/nixpkgs/issues/12847 and https://github.com/NixOS/nixpkgs/issues/67281.

Semi off-topic: have you tried using Geary instead?

**Comment by aanderse** [comment]
I may recall incorrectly but I think at one point this may have been on radar for @jtojnar or @worldofpeace after some other `gnome` work.

**Comment by itshaydendev** [comment]
@bjornfor Geary has no Exchange Server support. Only IMAP.

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    substituteInPlace CMakeLists.txt \
+    --replace "add_appdata_file" "#add_appdata_file"
+    substituteInPlace  src/addressbook/CMakeLists.txt \
+    --replace "e-book-backend-ews.c" ""
+    substituteInPlace  src/addressbook/CMakeLists.txt \
+    --replace "e-book-backend-ews-factory.c" ""
+  '';
+  
+  buildInputs = [
+    cmake gettext intltool pkgconfig evolution-data-server gnome3.evolution sqlite libgdata
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:24]: 1. Tools used for build like `cmake gettext intltool pkg-config` should go to `nativeBuildInputs`.
2. `pkg-config` should be used instead of the alias.

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    gtk3 webkitgtk
+  ];
+  
+  buildPhase = ''
+   rm -rf  src/addressbook/e-book-backend-ews.c
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:29]: Could you add a commend why would this be needed?

Also why are overriding `buildPhase`? This will prevent the build from running in there (will only get run as a dependency of `install` target). `preBuild` or `postPatch` hooks should be preferred for changes like this.

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    gtk3 webkitgtk
+  ];
+  
+  buildPhase = ''
+   rm -rf  src/addressbook/e-book-backend-ews.c 
+  '';
+  
+  cmakeFlags = [
+#    "-DCMAKE_INSTALL_PREFIX=${gnome3.evolution}"
+    "-DCMAKE_BUILD_TYPE=Release"
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:34]: CMake setup hook sets this by default:

https://github.com/NixOS/nixpkgs/blob/5bb1198cccb1a161e982dce84e18678ccec65529/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L92

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  
+   passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "evolution-ews";
+    };
+  };
+  
+  meta = with stdenv.lib; {
+    description = "Evolution connector for Microsoft Exchange Server protocols.";
+    license = licenses.lgpl2;
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:46]: This is fishy, opened an [issue](https://gitlab.gnome.org/GNOME/evolution-ews/-/issues/111) asking for clarification.

File: pkgs/desktops/gnome-3/default.nix [hunk 5]
```diff
@@ -42,6 +42,7 @@ lib.makeScope pkgs.newScope (self: with self; {
   evince = callPackage ./core/evince { }; # ToDo: dbus would prevent compilation, enable tests
 
   evolution-data-server = callPackage ./core/evolution-data-server { };
+  evolution-ews = callPackage ./misc/evolution-ews { };
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/default.nix:45]: Could you move this straight to `pkgs/applications/networking/mailreaders/`. I am slowly disbanding the `gnome3` package set and evolution will be moved to `all-packages.nix` next.

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 6]
```diff
@@ -0,0 +1,50 @@
+{ stdenv, fetchgit, gnome3, cmake, gettext, intltool, pkgconfig, evolution-data-server
+, sqlite, gtk3, webkitgtk, libgdata, ... }:
+
+stdenv.mkDerivation rec {
+  pname = "evolution-ews";
+  version = "3.36.3";
+  
+  src = fetchgit {
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:8]: Why not fetch it from GNOME mirrors?

https://github.com/NixOS/nixpkgs/blob/5bb1198cccb1a161e982dce84e18678ccec65529/pkgs/desktops/gnome-3/apps/evolution/default.nix#L49-L51

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    url = "https://gitlab.gnome.org/GNOME/${pname}.git/";
+    sha256 = "01qys33nddn27qbg9fk70v4kwkc80gszkfqnwi1grmi42d6igix0";
+    rev = "610facabb38bdf41d993ac0b21866f06fa1fc8dd"; # tag 3.36.3
+  };
+  
+  prePatch = ''
+    substituteInPlace CMakeLists.txt \
+    --replace "add_appdata_file" "#add_appdata_file"
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:16]: Why is this needed?

Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:16]: Why is this needed?

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix [hunk 8]
```diff
@@ -0,0 +1,50 @@
+{ stdenv, fetchgit, gnome3, cmake, gettext, intltool, pkgconfig, evolution-data-server
+, sqlite, gtk3, webkitgtk, libgdata, ... }:
```
Comment (jtojnar) [inline pkgs/desktops/gnome-3/misc/evolution-ews/default.nix:2]: Ellipsis not used. `callPackage` will only fill in the available arguments anyway.

```suggestion
, sqlite, gtk3, webkitgtk, libgdata }:
```

**Review by jtojnar — CHANGES_REQUESTED** [review]
Could you rebase (`git rebase -i master`, `git push -f`) the branch so that there is only the relevant commit and no useless merges?

**Comment by jtojnar** [comment]
The main issue is that evolution only looks for plug-ins in its directory tree, so we need to create `evolution-with-plugins` expression as described in https://github.com/NixOS/nixpkgs/issues/12847#issuecomment-354350825 or https://discourse.nixos.org/t/writing-derivation-for-a-plugin-to-evolution-data-server/2719

File: pkgs/desktops/gnome-3/default.nix (hunk 5 above)
Comment (dasj19) [reply pkgs/desktops/gnome-3/default.nix:45]: I don't quite get what you mean. Do I have to make an `pkgs/applications/networking/mailreaders/evolution` folder an a `evolution-ews.nix` file inside that?

Comment (jtojnar) [reply pkgs/desktops/gnome-3/default.nix:45]: I think `pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix` makes sense. Especially if we package other extensions in the future.

**Comment by dasj19** [comment]
I hit a roadblock:
```
CMake Error at src/addressbook/cmake_install.cmake:55 (file):
  file INSTALL cannot copy file
  "/build/evolution-ews-3.36.3/build/src/addressbook/libebookbackendews.so"
  to
  "/nix/store/0khik6f3gpz4gkvpjipy1g1hq7qbkknf-evolution-data-server-3.36.3/lib/evolution-data-server/addressbook-backends/libebookbackendews.so":
  Permission denied.
Call Stack (most recent call first):
  src/cmake_install.cmake:42 (include)
  cmake_install.cmake:55 (include)


make: *** [Makefile:82: install] Error 1
```

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix (hunk 6 above)
Comment (dasj19) [reply pkgs/desktops/gnome-3/misc/evolution-ews/default.nix]: done. now I'm fetching from gnome mirrors

File: pkgs/desktops/gnome-3/misc/evolution-ews/default.nix (hunk 2 above)
Comment (dasj19) [reply pkgs/desktops/gnome-3/misc/evolution-ews/default.nix]: fixed.

**Comment by jtojnar** [comment]
The paths come from here:

https://gitlab.gnome.org/GNOME/evolution-ews/-/blob/ad2d6373b8b6c3b83790ceb454256b6547256455/CMakeLists.txt#L206-213

You can override them using `PKG_CONFIG_$PACKAGE_$VARIABLE` environment variable, where `$PACKAGE` and `$VARIABLE` are uppercased names with non-alphanumeric characters replaced by underscores, for example `PKG_CONFIG_CAMEL_1_2_CAMEL_PROVIDERDIR` (see `pkg-config (1)`).

**Comment by jtojnar** [comment]
To be more precise, you can run something like `nix-shell -A evolution-ews --run "pkg-config camel-1.2 --variable camel_providerdir"`, which will spit out something like `/nix/store/7yds4g5hb0ig7zqg4nkc03ng07vhbn6a-evolution-data-server-3.36.3/lib/evolution-data-server/camel-providers` so you will put `PKG_CONFIG_CAMEL_1_2_CAMEL_PROVIDERDIR = "${placeholder "out"}/lib/evolution-data-server/camel-providers";` to the expression.

**Comment by dasj19** [comment]
Thanks a lot @jtojnar 
I did as you asked and added the following to the .nix expression
```
PKG_CONFIG_EVOLUTION_SHELL_3_0_ERRORDIR = "${placeholder "out"}/share/evolution/errors";
  PKG_CONFIG_EVOLUTION_SHELL_3_0_PRIVLIBDIR = "${placeholder "out"}/lib/evolution";
  PKG_CONFIG_CAMEL_1_2_CAMEL_PROVIDERDIR = "${placeholder "out"}/lib/evolution-data-server/camel-providers";
  PKG_CONFIG_LIBEDATABOOK_1_2_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/addressbook-backends";
  PKG_CONFIG_LIBEDATACAL_2_0_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/calendar-backends";
  PKG_CONFIG_LIBEBACKEND_1_2_MODULEDIR = "${placeholder "out"}/lib/evolution-data-server/registry-modules";
  PKG_CONFIG_EVOLUTION_SHELL_3_0_MODULEDIR = "${placeholder "out"}/lib/evolution/modules";
  PKG_CONFIG_EVOLUTION_DATA_SERVER_1_2_PRIVDATADIR = "${placeholder "out"}/share/evolution-data-server";
```
The thing is it does not help on that error at all. Maybe because there is an earlier error which says:
```
-   Found camel-1.2, version 3.36.3
Package libpcre was not found in the pkg-config search path.
Perhaps you should add the directory containing `libpcre.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libpcre', required by 'glib-2.0', not found
```
Sorry to bother that much but I don't have much clues nor experience to go further from here...

**Comment by dasj19** [comment]
There is some sparse documentation here as well: https://wiki.gnome.org/Apps/Evolution/Building#Build_evolution-ews

**Comment by jtojnar** [comment]
The pcre thing should just be a warning. If it still fails, the error is probably elsewhere.

**Comment by jtojnar** [comment]
Looks like you missed some underscores:

```patch
--- a/pkgs/applications/networking/mailreaders/evolution-ews/default.nix
+++ b/pkgs/applications/networking/mailreaders/evolution-ews/default.nix
@@ -13,8 +13,8 @@ stdenv.mkDerivation rec {
   PKG_CONFIG_EVOLUTION_SHELL_3_0_ERRORDIR = "${placeholder "out"}/share/evolution/errors";
   PKG_CONFIG_EVOLUTION_SHELL_3_0_PRIVLIBDIR = "${placeholder "out"}/lib/evolution";
   PKG_CONFIG_CAMEL_1_2_CAMEL_PROVIDERDIR = "${placeholder "out"}/lib/evolution-data-server/camel-providers";
-  PKG_CONFIG_LIBEDATABOOK_1_2_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/addressbook-backends";
-  PKG_CONFIG_LIBEDATACAL_2_0_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/calendar-backends";
+  PKG_CONFIG_LIBEDATA_BOOK_1_2_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/addressbook-backends";
+  PKG_CONFIG_LIBEDATA_CAL_2_0_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/calendar-backends";
   PKG_CONFIG_LIBEBACKEND_1_2_MODULEDIR = "${placeholder "out"}/lib/evolution-data-server/registry-modules";
   PKG_CONFIG_EVOLUTION_SHELL_3_0_MODULEDIR = "${placeholder "out"}/lib/evolution/modules";
   PKG_CONFIG_EVOLUTION_DATA_SERVER_1_2_PRIVDATADIR = "${placeholder "out"}/share/evolution-data-server";
```

**Comment by dasj19** [comment]
> Looks like you missed some underscores:
Nice found. Now it compiles, and I need to look into letting evolution read from the install path of evolution-ews.

**Comment by jtojnar** [comment]
Also you should add yourself to `maintainers/maintainer-list.nix` such that there are two commits:

* maintainers: add dasj19
* evolution-ews: init at 3.36.4

This should fix the CI error.

File: pkgs/desktops/gnome-3/default.nix (hunk 5 above)
Comment (dasj19) [reply pkgs/desktops/gnome-3/default.nix:45]: done, thanks for suggestion

File: maintainers/maintainer-list.nix [hunk 9]
```diff
@@ -1790,6 +1790,12 @@
     githubId = 4971975;
     name = "Janne Heß";
   };
+  dasj19 = {
+    email = "daniel@serbanescu.dk";
+    github = "dasj19";
+    githubId = 7589338;
+    name = "Daniel Șerbănescu";
+  }
```
Comment (jtojnar) [inline maintainers/maintainer-list.nix:1798]: missing semicolon

**Comment by dasj19** [comment]
> Could you rebase (`git rebase -i master`, `git push -f`) the branch so that there is only the relevant commit and no useless merges?

I finally managed to get the commit history down to one commit but because I had an update merge in between, I triggered automatically a review request from all the maintainers who merged things into master. Sorry for that, I'll try to not do it again, I still struggle to squash things...
I think I found an easier way that seems to work for me: https://stackoverflow.com/a/53336320/5073290

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+   passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "evolution-ews";
+    };
+  };
+
+  meta = with stdenv.lib; {
+    description = "Evolution connector for Microsoft Exchange Server protocols.";
+    license = licenses.lgpl2;
+    maintainers = [ "dasj19" ];
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:44]:
```suggestion
    maintainers = [ lib.maintainers.dasj19 ];
```

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+   passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "evolution-ews";
+    };
+  };
+
+  meta = with stdenv.lib; {
+    description = "Evolution connector for Microsoft Exchange Server protocols.";
+    license = licenses.lgpl2;
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:43]:
```suggestion
    license = "LGPL-2.1-only OR LGPL-3.0-only"; # https://gitlab.gnome.org/GNOME/evolution-ews/issues/111
```

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  buildInputs = [
+    evolution-data-server gnome3.evolution
+    sqlite libgdata
+    gtk3 webkitgtk
+  ];
+
+  cmakeFlags = [
+    "-DWITH_MSPACK=OFF"
+    "-DCMAKE_BUILD_TYPE=Debug"
+  ];
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:33]: It would be nicer to use https://github.com/NixOS/nixpkgs/blob/7b2bdbea75a202c416941798aeb9544e2596fa3d/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L92:

```suggestion
  ];

  cmakeBuildType = "Debug";
```

But perhaps we should use the default release builds.

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  };
+
+  PKG_CONFIG_EVOLUTION_SHELL_3_0_ERRORDIR = "${placeholder "out"}/share/evolution/errors";
+  PKG_CONFIG_EVOLUTION_SHELL_3_0_PRIVLIBDIR = "${placeholder "out"}/lib/evolution";
+  PKG_CONFIG_CAMEL_1_2_CAMEL_PROVIDERDIR = "${placeholder "out"}/lib/evolution-data-server/camel-providers";
+  PKG_CONFIG_LIBEDATA_BOOK_1_2_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/addressbook-backends";
+  PKG_CONFIG_LIBEDATA_CAL_2_0_BACKENDDIR = "${placeholder "out"}/lib/evolution-data-server/calendar-backends";
+  PKG_CONFIG_LIBEBACKEND_1_2_MODULEDIR = "${placeholder "out"}/lib/evolution-data-server/registry-modules";
+  PKG_CONFIG_EVOLUTION_SHELL_3_0_MODULEDIR = "${placeholder "out"}/lib/evolution/modules";
+  PKG_CONFIG_EVOLUTION_DATA_SERVER_1_2_PRIVDATADIR = "${placeholder "out"}/share/evolution-data-server";
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:20]: Could you move this under `cmakeFlags`, I believe it fits there better as it also configures the build.

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:20]: It does not build anymore if I move them to cmakeFlags. I end back here: https://github.com/NixOS/nixpkgs/pull/91504#issuecomment-653476825 when I try to move them to cmakeFlags

Comment (jtojnar) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:20]: Sorry, I meant moving them below/after `cmakeFlags` in the code, not inside.

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:20]: oh, ok that's doable ;)

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix]: done.

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+  cmakeBuildType = "Debug";
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:34]: Why do we keep this?

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix]: Done. Removed the debug line

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  nativeBuildInputs = [ cmake gettext intltool pkg-config ];
+
+  buildInputs = [
+    evolution-data-server gnome3.evolution
+    sqlite libgdata
+    gtk3 webkitgtk
+  ];
+
+  cmakeFlags = [
+    "-DWITH_MSPACK=OFF"
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:22]: Could you add a comment why this is disabled?

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix]: actually it is better to have mspack on according to https://wiki.gnome.org/Apps/Evolution/Building#Build_evolution-ews
I had it off to get it to work in the first place

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+  buildInputs = [
+    evolution-data-server gnome3.evolution
+    sqlite libgdata
+    gtk3 webkitgtk
+  ];
+
+  # Building with libmspack as reccommended: https://wiki.gnome.org/Apps/Evolution/Building#Build_evolution-ews
+  cmakeFlags = [
+    "-DWITH_MSPACK=ON"
```
Comment (jtojnar) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:23]: This should be on by default IIUC: https://gitlab.gnome.org/GNOME/evolution-ews/-/blob/ad2d6373b8b6c3b83790ceb454256b6547256455/CMakeLists.txt#L149

But I wonder how it can build without `libmspack` being added to deps.

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+  PKG_CONFIG_EVOLUTION_SHELL_3_0_MODULEDIR = "${placeholder "out"}/lib/evolution/modules";
+  PKG_CONFIG_EVOLUTION_DATA_SERVER_1_2_PRIVDATADIR = "${placeholder "out"}/share/evolution-data-server";
+
+   passthru = {
+    updateScript = gnome3.updateScript {
+      packageName = "evolution-ews";
+    };
+  };
+
+  meta = with stdenv.lib; {
```
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:41]: homepage?

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:41]: There is this Gnome Wiki https://wiki.gnome.org/Apps/Evolution/EWS but I don't think it is worth adding since it is more like a todo list for evolution-ews

Comment (worldofpeace) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:41]: the repo on gitlab would be fine too

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:41]: okay, done. repo addded to homage.

File: pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix (hunk 16 above)
Comment (jtojnar) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:23]: Right, it does not build. Could you add the dependency?

Comment (dasj19) [reply pkgs/applications/networking/mailreaders/evolution/evolution-ews/default.nix:24]: Done, dependency added

**Comment by jtojnar** [comment]
Thank you.

**Comment by aanderse** [comment]
Great news! If I'm on a `plasma` desktop running what steps do I take to try this out? Thank you all for your hard work! :tada:

**Comment by jtojnar** [comment]
We still need to patch evolution to allow loading plug-ins. Maybe joining the two derivations into a single one by recursive copying and fixing all the paths would be easiest for now.

**Comment by aanderse** [comment]
Great. Anything which allows me to stop using web mail :smile:

**Comment by dasj19** [comment]
I have tried to make a wrapper this morning and just making symlinks to the .so files in the /lib folder is not enough. Some other magic is needed. I'll continue reading through nix pills & the manual to get some more knowledge. I have to admit... there are not that many "with-plugins" packages one can get some inspiration from.

**Comment by dasj19** [comment]
I started up the work at https://github.com/NixOS/nixpkgs/pull/92775

**Comment by jtojnar** [comment]
Yeah, programs typically have hardcoded path to their libdir in the code so you would need to install the plug-ins to that path. That is not possible since store paths are immutable. One way to achieve that is to patch the source code to accept some writeable path like `/run/current-system/sw/lib/evolution-data-server`, or an environment variable that can be passed at runtime. (Example patches: [ofono](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/tools/networking/ofono/0001-Search-connectors-in-OFONO_PLUGIN_PATH.patch), [nautilus](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/desktops/gnome-3/core/nautilus/extension_dir.patch), [dleyna-core](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/development/libraries/dleyna-core/0001-Search-connectors-in-DLEYNA_CONNECTOR_PATH.patch), [gsignond](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/development/libraries/gsignond/plugin-load-env.patch); example postprocessing: [thunar](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/desktops/xfce/core/thunar/wrapper.nix), [gsignond](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/development/libraries/gsignond/wrapper.nix), [zathura](https://github.com/NixOS/nixpkgs/blob/141f9ce8073ac98205acaae18ea0f550fbba2b3d/pkgs/applications/misc/zathura/wrapper.nix)). But that might be too much work since there are like ten different plug-in directories, so maybe approach I mentioned [above](https://github.com/NixOS/nixpkgs/pull/91504#issuecomment-655816325) will be easier.

**Comment by dasj19** [comment]
Thanks for examples @jtojnar ... have you submitted any of those patches upstream? They seem to be nice features.

**Comment by jtojnar** [comment]
Some of them, yes. With mixed results. [dleyna-core](https://github.com/intel/dleyna-core/pull/50) is inactive, GDM [succeeded](https://gitlab.gnome.org/GNOME/gdm/-/merge_requests/27), with generic variable in the end. I also recall discussing xdg-dektop-portal and evolution with their respective authors (probably on IRC as I cannot find it) and I did not manage to convince them. Not sure I would even convince myself – environment variables are terrible mechanism. But unfortunately we do not have much options.
